pkg/fqdn/dnsproxy/proxy.go

```diff
@@ -901,32 +1072,32 @@ func shouldCompressResponse(request, response *dns.Msg) bool {
 	return false
 }
 
-func GetSelectorRegexMap(l7 policy.L7DataMap) (CachedSelectorREEntry, error) {
-	newRE := make(CachedSelectorREEntry)
-	for selector, l7Rules := range l7 {
-		if l7Rules == nil {
-			l7Rules = &policy.PerSelectorPolicy{L7Rules: api.L7Rules{DNS: []api.PortRuleDNS{{MatchPattern: "*"}}}}
+func GenerateRegexpAndFqdns(l7Rules *policy.PerSelectorPolicy) (string, []string) {
+	if l7Rules == nil {
+		l7Rules = &policy.PerSelectorPolicy{L7Rules: api.L7Rules{DNS: []api.PortRuleDNS{{MatchPattern: "*"}}}}
+	}
+	fqdnStrings := make([]string, 0, len(l7Rules.DNS)/2)
+	reStrings := make([]string, 0, len(l7Rules.DNS)/2)
+	for _, dnsRule := range l7Rules.DNS {
+		if len(dnsRule.MatchName) > 0 {
+			dnsRuleName := strings.ToLower(dns.Fqdn(dnsRule.MatchName))
+			fqdnStrings = append(fqdnStrings, dnsRuleName)
 		}
-		reStrings := make([]string, 0, len(l7Rules.DNS))
-		for _, dnsRule := range l7Rules.DNS {
-			if len(dnsRule.MatchName) > 0 {
-				dnsRuleName := strings.ToLower(dns.Fqdn(dnsRule.MatchName))
-				dnsPatternAsRE := matchpattern.ToRegexp(dnsRuleName)
-				reStrings = append(reStrings, "("+dnsPatternAsRE+")")
-			}
-			if len(dnsRule.MatchPattern) > 0 {
-				dnsPattern := matchpattern.Sanitize(dnsRule.MatchPattern)
+		if len(dnsRule.MatchPattern) > 0 {
+			dnsPattern := matchpattern.Sanitize(dnsRule.MatchPattern)
+			if strings.Contains(dnsPattern, "*") {
 				dnsPatternAsRE := matchpattern.ToRegexp(dnsPattern)
 				reStrings = append(reStrings, "("+dnsPatternAsRE+")")
+			} else {
+				fqdnStrings = append(fqdnStrings, dnsPattern)
 			}
 		}
-		mp := strings.Join(reStrings, "|")
-		rei, err := re.CompileRegex(mp)
-		if err != nil {
-			return nil, err
-		}
-		newRE[selector] = rei
+	}
+	if len(reStrings) == 0 {
+		// Short circuit to very fast non-matching regexp in case of no matchPatterns, where
+		// we do have some matchNames.
+		reStrings = append(reStrings, "a^")
```

Comment:
This seems strange to me. Could we instead make the logic detect if there's no matchpattern, then we skip checking against the regex? That way we can avoid this.

Reply:
Is it non-trivial? If it's trivial, then I'd say go for it in this PR.

Reply:
From the pure code perspective of this file, it should be quite straight forwards. My only concern is the restore dns rules functionality. In case we start serializing it as `null`, we "break" the old API, and we will have to solve edge those edge cases to ensure we don't break downgrades. I don't think I have time to tackle it this week, so will have to defer it to next week..

Reply:
We could follow up in another PR to properly solve this later also. For now we could keep the current regex trick.

Reply:
Hey! Thanks for taking a look @jrajahalme! 😄 

> Aren't we already breaking downgrades here? If non-regex names are saved by a new version, and the old version after downgrade ignores them it will not implement the same DNS policy after restore.

To be explicit, yes, the restored rule might not match a given domain, but the downgrade won't break due to invalid json input (eg. causing. nil ptr exception). I read the restore support as best-effort, but happy to make it map 1:1 if you think we should do that. In that case we can also write tests to make sure its water tight as well.


> To un-break the downgrade we may have to backport the restore code that can read the non-regex rules and check on them to earlier versions. The code to check for the restored rules is already separate (checkRestored()), so that should be doable. That change could be made to deal with null regex as well. These backports would not change the way rules are stored so there is no downgrade impact from them to older versions.

Yeah, fully on board to do that if you think thats the best approach, tho. we need to figure out the versioning strategy from my other comment to avoid people having null pointer dereferencing when downgrading.